File: app/src/main/java/com/drdisagree/iconify/services/TileNotchBarKiller.java
package com.drdisagree.iconify.services;

import android.service.quicksettings.Tile;
import android.service.quicksettings.TileService;

import com.drdisagree.iconify.R;
import com.drdisagree.iconify.config.Prefs;
import com.drdisagree.iconify.utils.overlay.OverlayUtil;

public class TileNotchBarKiller extends TileService {

    private boolean isNotchBarKillerEnabled = Prefs.getBoolean("IconifyComponentNBK.overlay");

    @Override
    public void onTileAdded() {
        super.onTileAdded();
    }

    @Override
    public void onStartListening() {
        super.onStartListening();
        Tile tile = getQsTile();
        tile.setState(isNotchBarKillerEnabled ? Tile.STATE_ACTIVE : Tile.STATE_INACTIVE);
        tile.updateTile();
    }

    @Override
    public void onStopListening() {
        super.onStopListening();
    }

    @Override
    public void onClick() {
        super.onClick();

        if (isNotchBarKillerEnabled) {
            OverlayUtil.disableOverlay("IconifyComponentNBK.overlay");
        } else {
            OverlayUtil.enableOverlay("IconifyComponentNBK.overlay");
        }

        isNotchBarKillerEnabled = !isNotchBarKillerEnabled;

        Tile tile = getQsTile();
        tile.setState(isNotchBarKillerEnabled ? Tile.STATE_ACTIVE : Tile.STATE_INACTIVE);
        tile.setLabel(getResources().getString(R.string.notch_bar_killer_title));
        tile.setSubtitle(isNotchBarKillerEnabled ? getResources().getString(R.string.general_on) : getResources().getString(R.string.general_off));
        tile.updateTile();
    }

    @Override
    public void onTileRemoved() {
        super.onTileRemoved();
    }
}

File: app/src/main/java/com/drdisagree/iconify/services/TileNotchBarKiller.kt
package com.drdisagree.iconify.services

import android.service.quicksettings.Tile
import android.service.quicksettings.TileService
import com.drdisagree.iconify.R
import com.drdisagree.iconify.common.Const.FRAMEWORK_PACKAGE
import com.drdisagree.iconify.common.Preferences.NOTCH_BAR_KILLER_SWITCH
import com.drdisagree.iconify.config.RPrefs.getBoolean
import com.drdisagree.iconify.config.RPrefs.putBoolean
import com.drdisagree.iconify.utils.SystemUtils
import com.drdisagree.iconify.utils.overlay.manager.resource.ResourceEntry
import com.drdisagree.iconify.utils.overlay.manager.resource.ResourceManager

class TileNotchBarKiller : TileService() {

    private var isNotchBarKillerEnabled = getBoolean(NOTCH_BAR_KILLER_SWITCH, false)

    override fun onStartListening() {
        super.onStartListening()

        val tile = qsTile
        tile.state = if (isNotchBarKillerEnabled) Tile.STATE_ACTIVE else Tile.STATE_INACTIVE
        tile.updateTile()
    }

    override fun onClick() {
        super.onClick()

        if (!SystemUtils.hasStoragePermission()) {
            val tile = qsTile
            tile.subtitle = resources.getString(R.string.need_storage_perm_title)
            tile.updateTile()
            return
        }

        isNotchBarKillerEnabled = !isNotchBarKillerEnabled

        putBoolean(NOTCH_BAR_KILLER_SWITCH, isNotchBarKillerEnabled)

        val tile = qsTile
        tile.state = if (isNotchBarKillerEnabled) Tile.STATE_ACTIVE else Tile.STATE_INACTIVE
        tile.label = resources.getString(R.string.notch_bar_killer_title)
        tile.subtitle =
            if (isNotchBarKillerEnabled) resources.getString(R.string.general_on) else resources.getString(
                R.string.general_off
            )
        tile.updateTile()

        if (isNotchBarKillerEnabled) {
            ResourceManager.buildOverlayWithResource(
                applicationContext,
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "bool",
                    "config_fillMainBuiltInDisplayCutout",
                    "true"
                ),
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "string",
                    "config_mainBuiltInDisplayCutout",
                    "M 0,0 L 0, 0 C 0,0 0,0 0,0"
                ),
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "string",
                    "config_mainBuiltInDisplayCutoutRectApproximation",
                    "@string/config_mainBuiltInDisplayCutout"
                )
            )
        } else {
            ResourceManager.removeResourceFromOverlay(
                applicationContext,
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "bool",
                    "config_fillMainBuiltInDisplayCutout"
                ),
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "string",
                    "config_mainBuiltInDisplayCutout"
                ),
                ResourceEntry(
                    FRAMEWORK_PACKAGE,
                    "string",
                    "config_mainBuiltInDisplayCutoutRectApproximation"
                )
            )
        }
    }
}
